ReSharper.FSharp/src/FSharp.Psi/src/CodeFormatter/FSharpCodeFormatter.cs

```diff
@@ -0,0 +1,137 @@
+using System.Collections.Concurrent;
+using JetBrains.ReSharper.Plugins.FSharp.Psi;
+using JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree;
+using JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CodeStyle;
+using JetBrains.ReSharper.Psi.ExtensionsAPI.Tree;
+using JetBrains.ReSharper.Psi.Impl.CodeStyle;
+using JetBrains.ReSharper.Psi.Parsing;
+using JetBrains.ReSharper.Psi.Tree;
+using JetBrains.ReSharper.Psi.Util;
+using JetBrains.Text;
+using JetBrains.Util.Text;
+using Whitespace = JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree.Whitespace;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Services.Formatter
+{
+  [Language(typeof(FSharpLanguage))]
+  public class FSharpCodeFormatter : CodeFormatterBase<FSharpFormatSettingsKey>
+  {
+    private readonly FSharpFormatterInfoProvider formatterInfoProvider;
+
+    private readonly ConcurrentDictionary<FormatterImplHelper.TokenTypePair, bool> glueingCache =
+      new ConcurrentDictionary<FormatterImplHelper.TokenTypePair, bool>();
+
+    public FSharpCodeFormatter(
+      FSharpLanguage language,
+      CodeFormatterRequirements requirements,
```

Comment:
Nit-picking: please don't put each arg on a separate line, such style is not used anywhere in the solution.